diff -ur d1x.orig/texmap/tmappent.S d1x/texmap/tmappent.S
--- d1x.orig/texmap/tmappent.S	1999-12-08 02:03:51.000000000 +0100
+++ d1x/texmap/tmappent.S	2005-08-07 11:28:04.000000000 +0200
@@ -110,7 +110,7 @@
 
 // This is used to abbreviate an annoying external variable name.
 
-.equ    fadetbl, _gr_fade_table
+#define    fadetbl _gr_fade_table
 
 
 // The following macro encapsulates the floating-point instructions
@@ -153,22 +153,22 @@
 .endm
 
 #ifdef __LINUX__
-.equ _pixptr, pixptr
-.equ _gr_fade_table, gr_fade_table
-.equ _write_buffer, write_buffer
-.equ _bytes_per_row,bytes_per_row
-.equ _fx_xleft, fx_xleft
-.equ _fx_xright, fx_xright
-.equ _fx_y, fx_y
-.equ _fx_u, fx_u
-.equ _fx_v, fx_v
-.equ _fx_z, fx_z
-.equ _fx_l, fx_l
-.equ _fx_du_dx, fx_du_dx
-.equ _fx_dv_dx, fx_dv_dx
-.equ _fx_dz_dx, fx_dz_dx
-.equ _fx_dl_dx, fx_dl_dx
-.equ _Transparency_on, Transparency_on
+#define _pixptr pixptr
+#define _gr_fade_table gr_fade_table
+#define _write_buffer write_buffer
+#define _bytes_per_row bytes_per_row
+#define _fx_xleft fx_xleft
+#define _fx_xright fx_xright
+#define _fx_y fx_y
+#define _fx_u fx_u
+#define _fx_v fx_v
+#define _fx_z fx_z
+#define _fx_l fx_l
+#define _fx_du_dx fx_du_dx
+#define _fx_dv_dx fx_dv_dx
+#define _fx_dz_dx fx_dz_dx
+#define _fx_dl_dx fx_dl_dx
+#define _Transparency_on Transparency_on
 
 .globl asm_pent_tmap_scanline_per
 #else
diff -ur d1x.orig/texmap/tmapppro.S d1x/texmap/tmapppro.S
--- d1x.orig/texmap/tmapppro.S	1999-12-08 02:03:51.000000000 +0100
+++ d1x/texmap/tmapppro.S	2005-08-07 11:27:08.000000000 +0200
@@ -109,7 +109,7 @@
 
 // This is used to abbreviate an annoying external variable name.
 
-.equ    fadetbl, _gr_fade_table
+#define fadetbl _gr_fade_table
 
 
 // The following macro encapsulates the floating-point instructions
@@ -121,7 +121,7 @@
 // dzdx to z, and finally puts 64 back onto the stack. At the end of
 // the macro, the stack contains: 64, z, u, v.
 
-.macro DoFPCalcs 0		// The FP stack after each instruction:
+.macro DoFPCalcs		// The FP stack after each instruction:
 				//		 64/z  z    u    v
 	fst	%st(4)		//		 64/z  z    u    v  64/z
 	fxch	%st(2)		//		   u   z  64/z   v  64/z
@@ -143,22 +143,22 @@
 
 
 #ifdef __LINUX__
-.equ _pixptr, pixptr
-.equ _gr_fade_table, gr_fade_table
-.equ _write_buffer, write_buffer
-.equ _bytes_per_row, bytes_per_row
-.equ _fx_xleft, fx_xleft
-.equ _fx_xright, fx_xright
-.equ _fx_y, fx_y
-.equ _fx_u, fx_u
-.equ _fx_v, fx_v
-.equ _fx_z, fx_z
-.equ _fx_l, fx_l
-.equ _fx_du_dx, fx_du_dx
-.equ _fx_dv_dx, fx_dv_dx
-.equ _fx_dz_dx, fx_dz_dx
-.equ _fx_dl_dx, fx_dl_dx
-.equ _Transparency_on, Transparency_on
+#define _pixptr pixptr
+#define _gr_fade_table gr_fade_table
+#define _write_buffer write_buffer
+#define _bytes_per_row bytes_per_row
+#define _fx_xleft fx_xleft
+#define _fx_xright fx_xright
+#define _fx_y fx_y
+#define _fx_u fx_u
+#define _fx_v fx_v
+#define _fx_z fx_z
+#define _fx_l fx_l
+#define _fx_du_dx fx_du_dx
+#define _fx_dv_dx fx_dv_dx
+#define _fx_dz_dx fx_dz_dx
+#define _fx_dl_dx fx_dl_dx
+#define _Transparency_on Transparency_on
 
 .globl asm_ppro_tmap_scanline_per
 #else
