--- d1x/rules.mak~	2006-09-26 08:47:44.000000000 +0200
+++ d1x/rules.mak	2006-09-26 08:47:44.000000000 +0200
@@ -162,7 +162,8 @@
 LINKLIBS += $(STATICSDL) -ldl -L/usr/X11R6/lib -lX11 -lXext -lXxf86dga -lXxf86vm
 CFLAGS += -D_SDL_STATIC_LIB
 else
-LINKLIBS += -ldl -lSDL
+# -lX11 to resolve the Xlib symbols used directly by the clipboard code
+LINKLIBS += -ldl -lSDL -lX11
 endif
 endif
 ifdef SVGALIB
