diff -up d1x/main/ipclient.cpp~ d1x/main/ipclient.cpp
--- d1x/main/ipclient.cpp~	2008-07-23 15:53:27.000000000 +0200
+++ d1x/main/ipclient.cpp	2008-07-23 15:53:27.000000000 +0200
@@ -2,6 +2,8 @@
  * ipclient.cpp - udp/ip client code
  * added 2000/02/07 Matt Mueller
  */
+#include <string.h> 
+
 extern "C"{
 #include "mono.h"
 #include "key.h"
diff -up d1x/main/ipserver.cpp~ d1x/main/ipserver.cpp
--- d1x/main/ipserver.cpp~	2008-07-23 15:56:10.000000000 +0200
+++ d1x/main/ipserver.cpp	2008-07-23 15:56:10.000000000 +0200
@@ -4,6 +4,7 @@
  */
 
 #include <stdlib.h>
+#include <string.h>
 #include <sys/types.h>
 #include <sys/time.h>
 #include <signal.h>
diff -up d1x/2d/bitblt.c~ d1x/2d/bitblt.c
--- d1x/2d/bitblt.c~	2008-07-23 16:02:36.000000000 +0200
+++ d1x/2d/bitblt.c	2008-07-23 16:02:36.000000000 +0200
@@ -1026,7 +1026,7 @@ void gr_bm_ubitblt0x_rle(int w, int h, i
 
 // rescalling bitmaps, 10/14/99 Jan Bobrowski jb@wizard.ae.krakow.pl
 
-inline void scale_line(byte *in, byte *out, int ilen, int olen)
+void scale_line(byte *in, byte *out, int ilen, int olen)
 {
 	int a = olen/ilen, b = olen%ilen;
 	int c = 0, i;
