diff -urbB d1x/main/kconfig.c ../d1x-rebirth_v0.42-src-20060416/main/kconfig.c
--- d1x/main/kconfig.c	2000-10-27 06:08:33.000000000 +0200
+++ ../d1x-rebirth_v0.42-src-20060416/main/kconfig.c	2006-03-26 12:51:56.000000000 +0200
@@ -1878,6 +1881,13 @@
 	ubyte channel_masks;
 	int use_mouse, use_joystick;
 	int speed_factor=1;
+	int ph_div;
+
+	if ((i=FindArg("-mouselook")) && (Config_control_type == 5) ) {
+		ph_div = 2;
+	} else {
+		ph_div = 8;
+	}
 
 	if (Game_turbo_mode)
 		speed_factor = 2;
@@ -2208,9 +2218,9 @@
 		//mprintf(( 0, "UM: %d, PV: %d\n", use_mouse, kc_mouse[13].value ));
 		if ( (use_mouse)&&(kc_mouse[13].value < 255) )	{
 			if ( !kc_mouse[14].value )		// If not inverted...
-				Controls.pitch_time -= (mouse_axis[kc_mouse[13].value]*Config_joystick_sensitivity)/8;
+				Controls.pitch_time -= (mouse_axis[kc_mouse[13].value]*Config_joystick_sensitivity)/ph_div; // ZICO we call pitch_div here for better pitching if mouselook
 			else
-				Controls.pitch_time += (mouse_axis[kc_mouse[13].value]*Config_joystick_sensitivity)/8;
+				Controls.pitch_time += (mouse_axis[kc_mouse[13].value]*Config_joystick_sensitivity)/ph_div;
 		}
 	} else {
 		Controls.pitch_time = 0;
@@ -2337,9 +2347,9 @@
 		// From mouse...
 		if ( (use_mouse)&&(kc_mouse[15].value < 255 ))	{
 			if ( !kc_mouse[16].value )		// If not inverted...
-				Controls.heading_time += (mouse_axis[kc_mouse[15].value]*Config_joystick_sensitivity)/8;
+				Controls.heading_time += (mouse_axis[kc_mouse[15].value]*Config_joystick_sensitivity)/ph_div;	// ZICO we call head_div here for better pitching if mouselook
 			else
-				Controls.heading_time -= (mouse_axis[kc_mouse[15].value]*Config_joystick_sensitivity)/8;
+				Controls.heading_time -= (mouse_axis[kc_mouse[15].value]*Config_joystick_sensitivity)/ph_div;
 		}
 	} else {
 		Controls.heading_time = 0;
@@ -2640,17 +2654,20 @@
 	if (FrameTime > F1_0 )
 		mprintf( (1, "Bogus frame time of %.2f seconds\n", f2fl(FrameTime) ));
 
+// ZICO - remove clamp for pitch and heading if mouselook on and no multiplayer game
+if ((Config_control_type != 5) || !(i=FindArg("-mouselook")) || (Game_mode & GM_MULTI) ) {
 	if (Controls.pitch_time > FrameTime/2 ) Controls.pitch_time = FrameTime/2;
-	if (Controls.vertical_thrust_time > FrameTime ) Controls.vertical_thrust_time = FrameTime;
 	if (Controls.heading_time > FrameTime ) Controls.heading_time = FrameTime;
+	if (Controls.pitch_time < -FrameTime/2 ) Controls.pitch_time = -FrameTime/2;
+	if (Controls.heading_time < -FrameTime ) Controls.heading_time = -FrameTime;
+}
+	if (Controls.vertical_thrust_time > FrameTime ) Controls.vertical_thrust_time = FrameTime;
 	if (Controls.sideways_thrust_time > FrameTime ) Controls.sideways_thrust_time = FrameTime;
 	if (Controls.bank_time > FrameTime ) Controls.bank_time = FrameTime;
 	if (Controls.forward_thrust_time > FrameTime ) Controls.forward_thrust_time = FrameTime;
 //	if (Controls.afterburner_time > FrameTime ) Controls.afterburner_time = FrameTime;
 
-	if (Controls.pitch_time < -FrameTime/2 ) Controls.pitch_time = -FrameTime/2;
 	if (Controls.vertical_thrust_time < -FrameTime ) Controls.vertical_thrust_time = -FrameTime;
-	if (Controls.heading_time < -FrameTime ) Controls.heading_time = -FrameTime;
 	if (Controls.sideways_thrust_time < -FrameTime ) Controls.sideways_thrust_time = -FrameTime;
 	if (Controls.bank_time < -FrameTime ) Controls.bank_time = -FrameTime;
 	if (Controls.forward_thrust_time < -FrameTime ) Controls.forward_thrust_time = -FrameTime;
