--- d1x/main/playsave.c.aap	2006-09-23 14:48:18.000000000 +0200
+++ d1x/main/playsave.c	2006-09-23 14:54:05.000000000 +0200
@@ -294,8 +294,6 @@
 	int	n_highest_levels;				//how many highest levels are saved
 	int	default_difficulty_level;
 	int	default_leveling_on;
-	int	Game_window_w;
-	int	Game_window_h;
 } __pack__ save_info;
 
 typedef struct hli {
@@ -335,6 +333,9 @@
 
 int Default_leveling_on=1;
 
+static int Player_game_window_w = 0;
+static int Player_game_window_h = 0;
+
 void init_game_list()
 {
 	int i;
@@ -668,6 +669,26 @@
                  }
                free(line);
              }
+            else if (strstr(word,"RESOLUTION"))
+             {
+               free(line); free(word);
+               line=fsplitword(f,'\n');
+               word=splitword(line,'=');
+               strupr(word);
+
+                while(!strstr(word,"END") && !feof(f))
+                 {
+                    if(!strcmp(word,"WIDTH"))
+                     sscanf(line,"%i",&Player_game_window_w);
+                    if(!strcmp(word,"HEIGHT"))
+                     sscanf(line,"%i",&Player_game_window_h);
+                   free(line); free(word);
+                   line=fsplitword(f,'\n');
+                   word=splitword(line,'=');
+                   strupr(word);
+                 }
+               free(line);
+             }
             else if (strstr(word,"END") || feof(f))
              {
               Stop=1;
@@ -966,6 +987,10 @@
           fprintf(fout,"0=0x%x,0x%x\n",kconfig_d1x_settings[18],kconfig_d1x_settings[19]);
           fprintf(fout,"[end]\n");
 //end this section change - VR
+          fprintf(fout,"[resolution]\n");
+          fprintf(fout,"width=%d\n", Game_window_w);
+          fprintf(fout,"height=%d\n", Game_window_h);
+          fprintf(fout,"[end]\n");
 
           fprintf(fout,"[plx version]\n");
           fprintf(fout,"plx version=%s\n",D1X_VERSION);
@@ -983,6 +1008,7 @@
 #define JOYSTICK 8
 #define NEWER_KEYS 16
 #define WEAPON_ORDER 32
+#define RESOLUTION 64
 
            while(!Stop && !feof(fin))
             {
@@ -1112,6 +1138,21 @@
                   free(line);
                   printed |= JOYSTICK;
                 }
+               else if (strstr(line,"RESOLUTION"))
+                {
+		  fprintf(fout,"[resolution]\n");
+		  fprintf(fout,"width=%d\n", Game_window_w);
+		  fprintf(fout,"height=%d\n", Game_window_h);
+		  fprintf(fout,"[end]\n");
+                   while(!strstr(line,"END")&&!feof(fin))
+                    {
+                      free(line);
+                      line=fsplitword(fin,'\n');
+                      strupr(line);
+                    }
+                  free(line);
+                  printed |= RESOLUTION;
+                }
                else if (strstr(line,"END"))
                 {
                   Stop=1;
@@ -1197,6 +1238,13 @@
               fprintf(fout,"[end]\n");
             }
 //end this section addition - VR
+           if(!(printed&RESOLUTION))
+            {
+	      fprintf(fout,"[resolution]\n");
+	      fprintf(fout,"width=%d\n", Game_window_w);
+	      fprintf(fout,"height=%d\n", Game_window_h);
+	      fprintf(fout,"[end]\n");
+	    }
 
           fprintf(fout,"[plx version]\n");
           fprintf(fout,"plx version=%s\n",D1X_VERSION);
@@ -1395,17 +1443,16 @@
 	}
 
 
-	if (Game_screen_mode != SM(info.Game_window_w,info.Game_window_h))
+	if (Player_game_window_w && Player_game_window_h &&
+		Game_screen_mode != SM(Player_game_window_w, Player_game_window_h))
 	{
 		/* Free VR_offscreen_buffer so that game_init_render_buffers
 		   recreates it. */
 		gr_free_canvas(VR_offscreen_buffer);
 		VR_offscreen_buffer = NULL;
-		Game_screen_mode = SM(info.Game_window_w,info.Game_window_h);
-		game_init_render_buffers(
-			SM(info.Game_window_w,info.Game_window_h),
-			info.Game_window_w,
-			info.Game_window_h, VR_NONE);
+		Game_screen_mode = SM(Player_game_window_w, Player_game_window_h);
+		game_init_render_buffers(Game_screen_mode,
+			Player_game_window_w, Player_game_window_h, VR_NONE);
 	}
 
 	return errno_ret;
@@ -1494,8 +1541,6 @@
 	info.player_struct_version = PLAYER_STRUCT_VERSION;
 	info.default_difficulty_level = Player_default_difficulty;
 	info.default_leveling_on = Auto_leveling_on;
-	info.Game_window_w = Game_window_w;
-	info.Game_window_h = Game_window_h;
 
 	info.n_highest_levels = n_highest_levels;
 
