--- d1x/main/inferno.c.help	2006-08-26 09:55:27.000000000 +0200
+++ d1x/main/inferno.c	2006-08-26 09:57:03.000000000 +0200
@@ -928,35 +928,24 @@
 
 void show_cmdline_help() {
 	printf( "%s\n", TXT_COMMAND_LINE_0 );
-        printf( "  -<X>x<Y>       %s\n", "Change screen resolution. Options:");
-        printf( "                    320x100;320x200;320x240;320x400;640x400;640x480;800x600;1024x768\n");
+	printf( "  -<X>x<Y>       %s\n",
+	  "Change screen resolution. Options: 320x100;320x200;320x240;");
+	printf( "                 %s\n",
+	  "320x400;640x400;640x480;800x600;1024x768");
 	printf( "%s\n", TXT_COMMAND_LINE_5 );
 	printf( "%s\n", TXT_COMMAND_LINE_6 );
-	printf( "%s\n", TXT_COMMAND_LINE_7 );
 	printf( "%s\n", TXT_COMMAND_LINE_8 );
-
-	printf( "\n%s\n",TXT_PRESS_ANY_KEY3);
-	getch();
-	printf( "\n" );
 	printf( "%s\n", TXT_COMMAND_LINE_9);
-	printf( "%s\n", TXT_COMMAND_LINE_10);
 	printf( "%s\n", TXT_COMMAND_LINE_11);
 	printf( "%s\n", TXT_COMMAND_LINE_12);
-	printf( "%s\n", TXT_COMMAND_LINE_13);
-	printf( "%s\n", TXT_COMMAND_LINE_14);
 	printf( "%s\n", TXT_COMMAND_LINE_15);
 	printf( "%s\n", TXT_COMMAND_LINE_16);
 	printf( "%s\n", TXT_COMMAND_LINE_17);
-	printf( "%s\n", TXT_COMMAND_LINE_18);
 	//printf( "  -DynamicSockets %s\n", TXT_SOCKET);
 	#ifdef __MSDOS__
 	printf( "  -NoFileCheck    %s\n", TXT_NOFILECHECK);
-	#endif
 	printf( "  -GamePort       %s\n", "Use Colorado Spectrum's Notebook Gameport" );
-//moved ifdef on 9/30/98 by Matt Mueller for linux doesn't use double buffering ever, so this is redundant
-	#ifdef __MSDOS__
 	printf( "  -NoDoubleBuffer %s\n", "Use only one page of video memory" );
-//end change -MM
 	printf( "  -JoyNice        %s\n", "Joystick poller allows interrupts to occur" );
 	#endif
 	printf( "\n%s\n",TXT_PRESS_ANY_KEY3);
@@ -979,10 +968,6 @@
         printf( "  -pilot <pilot>  %s\n", "Select this pilot automatically");
 	printf( "  -cockpit <n>    %s\n", "Set initial cockpit");
 	printf( "                  %s\n", "0=full 2=status bar 3=full screen");
-
-	printf( "\n%s\n",TXT_PRESS_ANY_KEY3);
-	getch();
-        printf( "\n\n");
 //added 02/07/99 Matt Mueller
         printf( "  -hud <h>        %s\n", "Set hud mode.  0=normal 1-3=new");
 //end addition -MM
@@ -994,7 +979,7 @@
         printf( "  -handicap <n>   %s\n", "Start game with <n> shields. Must be < 100 for multi");
         printf( "  -hudlog         %s\n", "Start hudlog immediatly");
 
-#ifdef SUPPORTS_NICEFPS
+#if 0 /* def SUPPORTS_NICEFPS, these default to on now */
         printf( "  -nicefps        %s\n", "Free cpu while waiting for next frame");
         printf( "  -niceautomap    %s\n", "Free cpu while doing automap");
 #endif
@@ -1002,11 +987,6 @@
         printf( "  -automap_gameres %s\n", "Set automap to use the same resolution as in game");
         printf( "  -menu<X>x<Y>    %s\n", "Set menu resolution to <X> by <Y>");
         printf( "  -menu_gameres   %s\n", "Set menus to use the same resolution as in game");
-
-	printf( "\n%s\n",TXT_PRESS_ANY_KEY3);
-	getch();
-        printf( "\n\n");
-
 //added/edited on 11/01/98 by Matthew Mueller
         printf( "  -hudlog_multi   %s\n", "Start hudlog upon entering multiplayer games");
         printf( "  -hudlogdir <d>  %s\n", "Log hud messages in directory <d>");
@@ -1024,10 +1004,13 @@
         printf( "  -font320 <f>    %s\n", "font to use for res 320x* and above (default font3-1.fnt)");
         printf( "  -font640 <f>    %s\n", "font to use for res 640x* and above (default pc6x8.fnt)");
         printf( "  -font800 <f>    %s\n", "font to use for res 800x* and above");
-        printf( "  -font1024 <f>    %s\n", "font to use for res 1024x* and above (default pc8x16.fnt)");
+        printf( "  -font1024 <f>   %s\n", "font to use for res 1024x* and above (default pc8x16.fnt)");
 		printf( "  -tmap <t>       %s\n","select texmapper to use (c,fp,i386,pent,ppro)");
-        printf( "\n\n");
-
+	printf( "  -mouselook      %s\n","Activate fast mouselook. Works in singleplayer only"); // ZICO - added for mouselook
+#ifdef __LINUX__
+	printf( "  -grabmouse      %s\n","Keeps the mouse from wandering out of the window"); // ZICO - added for mouse capture
+#endif
+	printf( "\n");
 	printf( "\n%s\n",TXT_PRESS_ANY_KEY3);
 	getch();
         printf( " System options:\n");
@@ -1036,13 +1019,16 @@
         printf( "  -joy209         %s\n", "Use alternate port 209 for joystick");
 #endif
 #ifdef GR_SUPPORTS_FULLSCREEN_TOGGLE 
-        printf( "  -fullscreen     %s\n", "Use fullscreen mode if available");
+	printf( "  -window         %s\n", "Run the game in a window"); // ZICO - from window to fullscreen
 #endif
 #ifdef OGL
+	printf( "  -16to9          %s\n", "use 16/9 widescreen");
+	printf( "  -16to10         %s\n", "use 16/10 widescreen");
         printf( "  -gl_texmaxfilt <f> %s\n","set GL_TEXTURE_MAX_FILTER (see readme.d1x)");
         printf( "  -gl_texminfilt <f> %s\n","set GL_TEXTURE_MIN_FILTER (see readme.d1x)");
         printf( "  -gl_mipmap      %s\n","set gl texture filters to \"standard\" options for mipmapping");
-        printf( "  -gl_simple      %s\n","set gl texture filters to gl_nearest for \"original\" look. (default)");
+	printf( "  -gl_trilinear   %s\n","set gl texture filters to trilinear mipmapping");
+        printf( "  -gl_simple      %s\n","set gl texture filters to gl_nearest for \"original\" look");
         printf( "  -gl_alttexmerge %s\n","use new texmerge, usually uses less ram (default)");
         printf( "  -gl_stdtexmerge %s\n","use old texmerge, uses more ram, but _might_ be a bit faster");
         printf( "  -gl_voodoo      %s\n","force fullscreen mode only");
@@ -1061,15 +1047,17 @@
 #ifdef SUPPORTS_NET_IP
         printf( "  -ip_nogetmyaddr %s\n","prevent autodetection of local ip address");
         printf( "  -ip_myaddr <a>  %s\n","use <a> as local ip address");
-        printf( "  -ip_baseport <p> %s\n","use <p> as offset from normal port (allows multiple instances of d1x to be run on a single computer)");
+        printf( "  -ip_baseport <p> %s\n","use <p> as offset from normal port");
 #endif
 #ifdef __LINUX__
 //added 02/06/99 Matthew Mueller
         printf( "  -serialdevice <s> %s\n", "Set serial/modem device to <s>");
 //end addition -MM
 //added 03/07/99 Matthew Mueller
-        printf( "  -serialread <r>   %s\n", "Set serial/modem to read from <r>");
+        printf( "  -serialread <r> %s\n", "Set serial/modem to read from <r>");
 //end addition -MM
+	printf( "  -playmus        %s\n", "Use playmus to play music (default)");
+	printf( "  -timidity       %s\n", "Use timidity to play music (better but much slower)");
 #endif
         printf( "\n");
 }
