diff -up d1x-rebirth_v0.57.1-src/SConstruct~ d1x-rebirth_v0.57.1-src/SConstruct
--- d1x-rebirth_v0.57.1-src/SConstruct~	2011-07-19 11:40:25.000000000 +0200
+++ d1x-rebirth_v0.57.1-src/SConstruct	2011-09-15 20:00:39.362419013 +0200
@@ -278,7 +278,7 @@ env.Append(CPPFLAGS = ['-Wall', '-funsig
 env.Append(CPPDEFINES = [('PROGRAM_NAME', '\\"' + str(PROGRAM_NAME) + '\\"'), ('D1XMAJOR', '\\"' + str(D1XMAJOR) + '\\"'), ('D1XMINOR', '\\"' + str(D1XMINOR) + '\\"'), ('D1XMICRO', '\\"' + str(D1XMICRO) + '\\"')])
 env.Append(CPPDEFINES = ['NETWORK', '_REENTRANT'])
 env.Append(CPPPATH = ['include', 'main', 'arch/include'])
-libs = ['physfs']
+libs = ['physfs', 'm']
 
 # Get traditional compiler environment variables
 if os.environ.has_key('CC'):
