diff -up d1x/main/game.c.foo d1x/main/game.c
--- d1x/main/game.c.foo	2008-07-23 15:37:03.000000000 +0200
+++ d1x/main/game.c	2008-07-23 15:37:17.000000000 +0200
@@ -4026,16 +4026,8 @@ void ReadControls()
 					Function_mode=FMODE_EXIT;
 					break;
 
-#ifdef SHAREWARE
-				case KEY_ALTED+KEY_F2:
-				case KEY_ALTED+KEY_F3:
-					hud_message( MSGC_GAME_FEEDBACK, TXT_ONLY_REGISTERED );
-					digi_play_sample( SOUND_BAD_SELECTION, F1_0 );
-					break;
-#else
 				case KEY_ALTED+KEY_F2:	state_save_all( 0 );		break;	// 0 means not between levels.
 				case KEY_ALTED+KEY_F3:	state_restore_all(1);		break;
-#endif
 
                                 //use function keys for window sizing
 
diff -up d1x/main/menu.c.foo d1x/main/menu.c
--- d1x/main/menu.c.foo	2008-07-23 15:37:03.000000000 +0200
+++ d1x/main/menu.c	2008-07-23 15:37:17.000000000 +0200
@@ -228,11 +228,6 @@ void create_main_menu(newmenu_item *m, i
 	
 
 	ADD_ITEM(TXT_NEW_GAME,MENU_NEW_GAME,KEY_N);
-
-#ifdef SHAREWARE
-	if (get_game_list(NULL)>0)
-#endif
-
   	ADD_ITEM(TXT_LOAD_GAME,MENU_LOAD_GAME,KEY_L);
 
 #ifdef NETWORK
@@ -342,11 +337,7 @@ void do_option ( int select) 
 			}
 			break;
 		case MENU_LOAD_GAME:
-#ifdef SHAREWARE
-			do_load_game_menu();
-#else
 			state_restore_all(0);	
-#endif
 			break;
 		#ifdef EDITOR
 		case MENU_EDITOR:
diff -up d1x/main/playsave.c.foo d1x/main/playsave.c
--- d1x/main/playsave.c.foo	2008-07-23 15:37:04.000000000 +0200
+++ d1x/main/playsave.c	2008-07-23 15:37:17.000000000 +0200
@@ -1474,9 +1474,8 @@ int read_player_file()
 	if (fclose(file) && errno_ret==EZERO)
 		errno_ret = errno;
 
-#ifndef SHAREWARE
-	if ( info.saved_game_version == COMPATIBLE_SAVED_GAME_VERSION ) 	{
-		int i;
+	if ( info.saved_game_version != 7 ) 	{
+		int i, found = 0;
 		
 		Assert( N_SAVE_SLOTS == 10 );
 
@@ -1485,11 +1484,15 @@ int read_player_file()
 				state_save_old_game(i, saved_games[i].name, &saved_games[i].player, 
              		saved_games[i].difficulty_level, saved_games[i].primary_weapon, 
           			saved_games[i].secondary_weapon, saved_games[i].next_level_num );
+          			// make sure we do not do this again, which would possibly overwrite
+          			// a new newstyle savegame
+          			saved_games[i].name[0] = 0;
+          			found++;
 			}
 		}
-		write_player_file();
+		if (found)
+			write_player_file();
 	}
-#endif
 
 	filename[strlen(filename) - 4] = 0;
 	strcat(filename, ".plx");
diff -up d1x/main/state.c.foo d1x/main/state.c
--- d1x/main/state.c.foo	2000-10-27 01:58:45.000000000 +0200
+++ d1x/main/state.c	2008-07-23 15:37:17.000000000 +0200
@@ -274,8 +274,6 @@ static char rcsid[] = "$Id: state.c,v 1.
 #include "gamefont.h"
 
 
-#ifndef SHAREWARE
-
 #define STATE_VERSION 7
 #define STATE_COMPATIBLE_VERSION 6
 // 0 - Put DGSS (Descent Game State Save) id at tof.
@@ -348,6 +346,7 @@ void state_callback(int nitems,newmenu_i
 //	}	
 }
 
+#if 0
 void rpad_string( char * string, int max_chars )
 {
 	int i, end_found;
@@ -362,6 +361,7 @@ void rpad_string( char * string, int max
 	}
 	*string = 0;		// NULL terminate
 }
+#endif
 
 int state_get_save_file(char * fname, char * dsc, int multi )
 {
@@ -577,8 +577,11 @@ int state_save_old_game(int slotnum, cha
 	fwrite( &temp_int, sizeof(int), 1, fp );
 
 // Save the mission info...
+#ifndef SHAREWARE
 	fwrite( &Mission_list[0], sizeof(char)*9, 1, fp );
-
+#else
+	fwrite( "\0\0\0\0\0\0\0\0", sizeof(char)*9, 1, fp );
+#endif
 //Save level info
 	temp_int = sg_player->level;
 	fwrite( &temp_int, sizeof(int), 1, fp );
@@ -625,6 +628,7 @@ int state_save_all(int between_levels)
 {
 	char filename[128], desc[DESC_LENGTH+1];
 
+#ifndef SHAREWARE
 	if ( Game_mode & GM_MULTI )	{
 #ifdef MULTI_SAVE
 		if ( FindArg( "-multisave" ) )
@@ -634,6 +638,7 @@ int state_save_all(int between_levels)
 			hud_message( MSGC_GAME_FEEDBACK, "Can't save in a multiplayer game!" );
 		return 0;
 	}
+#endif
 
 	mprintf(( 0, "CL=%d, NL=%d\n", Current_level_num, Next_level_num ));
 	
@@ -653,12 +658,14 @@ int state_save_all_sub(char *filename, c
 	FILE * fp;
 	grs_canvas * cnv;
 
+#ifndef SHAREWARE
 	if ( Game_mode & GM_MULTI )	{
 #ifdef MULTI_SAVE
 		if ( !FindArg( "-multisave" ) ) 
 #endif  
 			return 0;
 	}
+#endif
 
 	fp = fopen( filename, "wb" );
 	if ( !fp ) {
@@ -721,7 +728,11 @@ int state_save_all_sub(char *filename, c
 	fwrite( &between_levels, sizeof(int), 1, fp );
 
 // Save the mission info...
+#ifndef SHAREWARE
 	fwrite( &Mission_list[Current_mission_num], sizeof(char)*9, 1, fp );
+#else
+	fwrite( "\0\0\0\0\0\0\0\0", sizeof(char)*9, 1, fp );
+#endif
 
 //Save level info
 	fwrite( &Current_level_num, sizeof(int), 1, fp );
@@ -921,15 +932,32 @@ int state_restore_all_sub(char *filename
 // Read the mission info...
 	fread( mission, sizeof(char)*9, 1, fp );
 
+#ifndef SHAREWARE
 	if (!load_mission_by_name( mission ))	{
 		nm_messagebox( NULL, 1, "Ok", "Error!\nUnable to load mission\n'%s'\n", mission );
 		fclose(fp);
 		return 0;
 	}
+#else
+	if (mission[0]) {
+		nm_messagebox( NULL, 1, "Ok", "Error!\nCannot load mission '%s'\nThe shareware version only supports savegames of the shareware mission!", mission );
+		fclose(fp);
+		return 0;
+	}
+#endif
 
 //Read level info
 	fread( &current_level, sizeof(int), 1, fp );
 	fread( &next_level, sizeof(int), 1, fp );
+#ifdef SHAREWARE
+	if (current_level < 1 || current_level > Last_level ||
+		next_level < 0 || next_level > Last_level) {
+		nm_messagebox( NULL, 1, "Ok", "Error!\nCannot load level %d\nThe shareware version only supports savegames of shareware levels!",
+			between_levels? next_level:current_level);
+		fclose(fp);
+		return 0;
+	}
+#endif
 
 //Restore GameTime
 	fread( &GameTime, sizeof(fix), 1, fp );
@@ -1127,6 +1155,3 @@ RetryObjectLoading:
 
 	return 1;
 }
-
-
-#endif
